test/jdk/java/nio/channels/DatagramChannel/AdaptorAsyncCloseAfterReceive.java

```diff
@@ -0,0 +1,115 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8280113
+ * @summary Test async close of a DatagramSocket obtained from a DatagramChannel where
+ *     the DatagramChannel's internal socket address caches are already populated
+ * @enablePreview
+ * @library /test/lib
+ * @run junit AdaptorAsyncCloseAfterReceive
+ */
+
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketException;
+import java.nio.ByteBuffer;
+import java.nio.channels.DatagramChannel;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import jdk.test.lib.thread.VThreadRunner;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.AfterAll;
+import static org.junit.jupiter.api.Assertions.*;
+
+class AdaptorAsyncCloseAfterReceive {
+
+    // used for scheduling socket close
+    private static ScheduledExecutorService scheduler;
+
+    @BeforeAll
+    static void setup() {
+        scheduler = Executors.newSingleThreadScheduledExecutor();
+    }
+
+    @AfterAll
+    static void finish() {
+        scheduler.shutdown();
+    }
+
+    /**
+     * Test closing a DatagramSocket, obtained from a DatagramChannel, while the main
+     * thread is blocked in receive. The receive method should throw rather than
+     * completing with the sender address of a previous datagram.
+     */
+    @ParameterizedTest
+    @CsvSource({"0,0", "100,0", "0,60000", "100,60000"})
+    void testReceive(int maxLength, int timeout) throws Exception {
+        try (DatagramChannel dc = DatagramChannel.open()) {
+            dc.bind(new InetSocketAddress(InetAddress.getLoopbackAddress(), 0));
+
+            populateSocketAddressCaches(dc);
+
+            // schedule socket to be closed while main thead blocked in receive
+            DatagramSocket s = dc.socket();
+            Future<?> future = scheduler.schedule(() -> s.close(), 1, TimeUnit.SECONDS);
```

Comment:
Hello Alan, would it be a good idea to schedule this close as late as possible, just before the call to `s.receive(p)`? On heavily loaded systems, that would then prevent potential intermittent failures where this `s.close()` task could get fired before the control reaches `s.setSoTimeout(timeout)` and the `setSoTimeout(...)` thus would fail with a `SocketException`.

Reply:
That's a good point, it should be after the setSoTimeout to avoid an intermittent failure on a very slow/overloaded test machine.